# Scikit-learn exploration

## Made by:
- Annisa Rahim / 13518089
- Stefanus Gusega Gunawan / 13518149

# Breast-Cancer Dataset

Load the datasets first.

In [87]:
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

load = load_breast_cancer()
cancer = pd.DataFrame(load.data, columns=load.feature_names)
target = pd.DataFrame(load.target, columns =['target'])

# Split 80 : 20
X_train, X_valid, y_train, y_valid = train_test_split(cancer, target, train_size = 0.8, test_size = 0.2, random_state=1)
X_train

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
408        17.99         20.66          117.80      991.7          0.10360   
4          20.29         14.34          135.10     1297.0          0.10030   
307         9.00         14.40           56.36      246.3          0.07005   
386        12.21         14.09           78.78      462.0          0.08108   
404        12.34         14.95           78.29      469.1          0.08682   
434        14.86         16.94           94.89      673.7          0.08924   
19         13.54         14.36           87.46      566.3          0.09779   
517        19.89         20.26          130.50     1214.0          0.10370   
535        20.55         20.86          137.80     1308.0          0.10460   
445        11.99         24.89           77.61      441.3          0.10300   
554        12.88         28.92           82.50      514.3          0.08123   
236        23.21         26.97          153.50     1670.0          0.09509   
117        14.87         16.67           98.64      682.5          0.11620   
157        16.84         19.46          108.40      880.2          0.07445   
162        19.59         18.15          130.70     1214.0          0.11200   
78         20.18         23.97          143.70     1245.0          0.12860   
409        12.27         17.92           78.41      466.1          0.08685   
484        15.73         11.28          102.80      747.2          0.10430   
334        12.30         19.02           77.88      464.4          0.08313   
42         19.07         24.81          128.30     1104.0          0.09081   
173        11.08         14.71           70.21      372.7          0.10060   
223        15.75         20.25          102.60      761.3          0.10250   
201        17.54         19.32          115.10      951.6          0.08968   
133        15.71         13.93          102.00      761.7          0.09462   
232        11.22         33.81           70.79      386.8          0.07780   
413        14.99         22.11           97.53      693.7          0.08515   
514        15.05         19.07           97.26      701.9          0.09215   
244        19.40         23.50          129.10     1155.0          0.10270   
415        11.89         21.17           76.39      433.8          0.09773   
562        15.22         30.62          103.40      716.9          0.10480   
..           ...           ...             ...        ...              ...   
264        17.19         22.07          111.60      928.3          0.09726   
209        15.27         12.91           98.17      725.5          0.08182   
316        12.18         14.08           77.25      461.4          0.07734   
513        14.58         13.66           94.29      658.8          0.09832   
313        11.54         10.72           73.73      409.1          0.08597   
534        10.96         17.62           70.79      365.6          0.09687   
319        12.43         17.00           78.60      477.3          0.07557   
7          13.71         20.83           90.20      577.9          0.11890   
393        21.61         22.28          144.40     1407.0          0.11670   
141        16.11         18.05          105.10      813.0          0.09721   
86         14.48         21.46           94.25      648.2          0.09444   
478        11.49         14.59           73.99      404.9          0.10460   
503        23.09         19.83          152.10     1682.0          0.09342   
215        13.86         16.93           90.96      578.9          0.10260   
398        11.06         14.83           70.31      378.2          0.07741   
490        12.25         22.44           78.18      466.5          0.08192   
252        19.73         19.82          130.70     1206.0          0.10620   
468        17.60         23.33          119.00      980.5          0.09289   
357        13.87         16.21           88.52      593.7          0.08743   
254        19.45         19

Check whether there is missing values.

In [88]:
cols_missing = [col for col in cancer.columns if (cancer[col].isnull().any())]
cols_missing

[]

## Train datasets
### DecisionTreeClassifier

In [89]:
from sklearn.tree import DecisionTreeClassifier

classifier = DecisionTreeClassifier(random_state=1)
classifier.fit(X_train, y_train)

DecisionTreeClassifier(random_state=1)

Print the tree.

In [90]:
from sklearn.tree import export_text

tree = export_text(classifier, feature_names = list(cancer.columns))
print(tree)

|--- worst perimeter <= 106.05
|   |--- worst concave points <= 0.16
|   |   |--- worst concave points <= 0.14
|   |   |   |--- area error <= 48.98
|   |   |   |   |--- class: 1
|   |   |   |--- area error >  48.98
|   |   |   |   |--- worst radius <= 13.55
|   |   |   |   |   |--- class: 0
|   |   |   |   |--- worst radius >  13.55
|   |   |   |   |   |--- class: 1
|   |   |--- worst concave points >  0.14
|   |   |   |--- worst texture <= 29.45
|   |   |   |   |--- class: 1
|   |   |   |--- worst texture >  29.45
|   |   |   |   |--- class: 0
|   |--- worst concave points >  0.16
|   |   |--- worst texture <= 24.78
|   |   |   |--- class: 1
|   |   |--- worst texture >  24.78
|   |   |   |--- class: 0
|--- worst perimeter >  106.05
|   |--- worst texture <= 20.65
|   |   |--- worst perimeter <= 116.80
|   |   |   |--- class: 1
|   |   |--- worst perimeter >  116.80
|   |   |   |--- mean smoothness <= 0.08
|   |   |   |   |--- class: 1
|   |   |   |--- mean smoothness >  0.08
|   |   

Give prediction to training data so we can see how good the model is. Compared to true value of training data. 

In [91]:
predict_train = classifier.predict(X_train)

Evaluate the model with Accuracy Metrics.

$$
Accuracy = \frac {Number of Correct predictions}{Total number of predictions made}
$$

In [92]:
from sklearn.metrics import accuracy_score

acc_train_score = accuracy_score(y_train, predict_train)
acc_train_score

1.0

Evaluate the model with F1 score Metrics. The metrics used based on this confusion matrix. F1 Score is the **_Harmonic Mean_** between precision and recall.
<!-- ![Alt Text](image path "title") -->
![Alt Text](https://miro.medium.com/max/700/1*OhEnS-T54Cz0YSTl_c3Dwg.jpeg "Confusion Matrix")

$$
Precision = \frac {TruePositives}{TruePositives+FalsePositives}
$$ <br>
$$
Recall = \frac {TruePositives}{TruePositives+FalseNegatives}
$$ <br>
$$
F1 = 2*(\frac{1}{\frac {1}{Precision} + \frac {1}{Recall}}) = 2*(\frac {TruePositives}{2TruePositives+FalsePositives+FalseNegatives})
$$

In [93]:
from sklearn.metrics import f1_score

f1_train_score = f1_score(y_train, predict_train)
f1_train_score

1.0

Now predict the test data.

In [94]:
predict_valid = classifier.predict(X_valid)

Evaluate the model with Accuracy Metrics.

In [95]:
acc_valid_score = accuracy_score(y_valid, predict_valid)
acc_valid_score

0.9473684210526315

Evaluate the model with F1 Metrics

In [96]:
f1_valid_score = f1_score(y_valid, predict_valid)
f1_valid_score

0.9594594594594595

### Id3Estimator

Train the model and print the tree.

In [97]:
import six
import sys
sys.modules['sklearn.externals.six'] = six

from id3 import Id3Estimator, export

id3_model1 = Id3Estimator()

id3_model1.fit(X_train, y_train.values.ravel()) # make it numpy 1D array, not a df

# export text
tree_text1 = export.export_text(id3_model1.tree_, feature_names=cancer.columns)

print(tree_text1)



worst perimeter <=105.15
|   worst concave points <=0.14
|   |   radius error <=0.64: 1 (249) 
|   |   radius error >0.64
|   |   |   mean radius <=12.27: 0 (1) 
|   |   |   mean radius >12.27: 1 (2) 
|   worst concave points >0.14
|   |   worst texture <=25.94: 1 (6) 
|   |   worst texture >25.94
|   |   |   mean compactness <=0.12
|   |   |   |   mean radius <=14.06: 1 (2) 
|   |   |   |   mean radius >14.06: 0 (1) 
|   |   |   mean compactness >0.12: 0 (5) 
worst perimeter >105.15
|   worst concave points <=0.15
|   |   worst texture <=19.91: 1 (13) 
|   |   worst texture >19.91
|   |   |   worst radius <=16.80
|   |   |   |   mean smoothness <=0.09: 1 (8) 
|   |   |   |   mean smoothness >0.09
|   |   |   |   |   smoothness error <=0.00: 1 (2) 
|   |   |   |   |   smoothness error >0.00: 0 (5) 
|   |   |   worst radius >16.80
|   |   |   |   worst concavity <=0.21
|   |   |   |   |   mean texture <=21.26: 1 (2) 
|   |   |   |   |   mean texture >21.26: 0 (2) 
|   |   |   |   worst

Evaluate the model with Accuracy and F1 score metrics.

In [98]:
# Model Evaluation
predict_train1 = id3_model1.predict(X_train)

print("Accuracy: ", accuracy_score(predict_train1, y_train))
print("F1 score: ", f1_score(predict_train1, y_train))

Accuracy:  1.0
F1 score:  1.0


Evaluate the model to test data with Accuracy and F1 score metrics.

In [99]:
# Model Prediction
predict_test1 = id3_model1.predict(X_valid)

print("Accuracy: ", accuracy_score(predict_test1,y_valid))
print("F1 score: ", f1_score(predict_test1,y_valid))

Accuracy:  0.9385964912280702
F1 score:  0.953020134228188


### KMeans

Train the datasets.

In [100]:
from sklearn.cluster import KMeans

# Determine how many clusters
n_clust = y_train['target'].nunique()

kmeans_model_1 = KMeans(n_clusters=n_clust,random_state=1)
kmeans_model_1.fit(X_train)

KMeans(n_clusters=2, random_state=1)

Predict the data train and data test.

In [101]:
kmeans_predict_train = kmeans_model_1.predict(X_train)
kmeans_predict_test = kmeans_model_1.predict(X_valid)
print(kmeans_predict_train)
print(y_train)

[1 1 0 0 0 0 0 1 1 0 0 1 0 0 1 1 0 0 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0
 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0 1
 0 0 1 1 0 0 0 0 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1 1 1 1 0 1 0 0 0 0 0 0 0 0
 0 0 0 0 1 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 1 0 0 1 1 0
 0 0 1 0 0 0 0 1 1 0 0 0 1 0 0 0 1 0 0 1 0 0 0 0 1 0 0 1 1 0 0 0 1 1 0 0 1
 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 1 1 0
 0 0 0 1 0 0 1 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 0 0 0 0 0 0 1
 1 1 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 1
 1 1 0 0 0 1 1 0 0 0 0 1 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 1 1 0 0 0 0 0 0 1 0
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 1 1 0
 1 0 0 0 0 0 1 0 1 0 0]
     target
408       0
4         0
307       1
386       1
404       1
434 

Define a function to switch the value from 0 to 1

In [102]:
import numpy as np
def switch(x: int) -> int:
    if (x==0):
        return x+1
    elif (x==1):
        return x-1

For evaluation with training data.

In [103]:
# Find out each cluster belong with which class

# First clusterization using first prediction (k_means_train)
# Calculate the accuracy and F1
acc_train = accuracy_score(y_train, kmeans_predict_train)
f1_train = f1_score(y_train, kmeans_predict_train)
print(f"First clusterization accuracy: {acc_train}")
print(f"First clusterization F1: {f1_train}")

# Second clusterization: switch 0 to 1
switched_train = pd.Series(kmeans_predict_train)
fix = switched_train.apply(switch)
acc_train1= accuracy_score(y_train, fix)
f1_train2= f1_score(y_train, fix)
print(f"Second clusterization accuracy: {acc_train1}")
print(f"Second clusterization F1: {f1_train2}")

First clusterization accuracy: 0.15164835164835164
First clusterization F1: 0.005154639175257732
Second clusterization accuracy: 0.8483516483516483
Second clusterization F1: 0.891679748822606


The performance can be seen by the greatest metrics score. So the cluster 0 belongs to class of 1, cluster 1 belongs to class of 0. Now, for evaluation with validation data (datatest).

In [104]:
# First clusterization
acc_valid = accuracy_score(y_valid, kmeans_predict_test)
f1_valid = f1_score(y_valid, kmeans_predict_test)
print(f"First clusterization accuracy: {acc_valid}")
print(f"First clusterization F1: {f1_train}")

# Second clusterization: switch 0 to 1
switched_valid = pd.Series(kmeans_predict_test).apply(switch)
acc_valid_switch = accuracy_score(y_valid, switched_valid)
f1_valid_switch = f1_score(y_valid, switched_valid)
print(f"Second clusterization accuracy: {acc_valid_switch}")
print(f"Second clusterization F1: {f1_valid_switch}")

First clusterization accuracy: 0.19298245614035087
First clusterization F1: 0.005154639175257732
Second clusterization accuracy: 0.8070175438596491
Second clusterization F1: 0.8674698795180723


Same as when we evaluate with training data, the second clusterization is better.

### LogisticRegression

In [105]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score

model = LogisticRegression()

model.fit(X_train,y_train.values.ravel())

c:\users\annisa rahim\appdata\local\programs\python\python37-32\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression()

In [106]:
predict_train = model.predict(X_train)

acc = accuracy_score(y_train, predict_train)
f1 = f1_score(y_train, predict_train)

print("Local Data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

Local Data
Accuracy:  0.945054945054945
F1 score:  0.9566724436741767


In [107]:
predict_test = model.predict(X_valid)

acc = accuracy_score(y_valid, predict_test)
f1 = f1_score(y_valid, predict_test)

print("Test Data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

Test Data
Accuracy:  0.9473684210526315
F1 score:  0.9589041095890412


Scaling the dataset (using Standard Scaler)

In [108]:
# Scaling untuk mencapai konvergensi dari data (menghindari ConvergenceWarning)
from sklearn.preprocessing import StandardScaler, MinMaxScaler
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_valid = scaler.fit_transform(X_valid)

Modelling the scaled dataset.

In [109]:
model.fit(X_train,y_train.values.ravel())

LogisticRegression()

Evaluate the model to train data with Accuracy and F1 score metrics.

In [110]:
predict_train = model.predict(X_train)

acc = accuracy_score(y_train, predict_train)
f1 = f1_score(y_train, predict_train)

print("Local Data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

Local Data
Accuracy:  0.9912087912087912
F1 score:  0.993006993006993


Evaluate the model to test data with Accuracy and F1 score metrics.

In [111]:
predict_test = model.predict(X_valid)

acc = accuracy_score(y_valid, predict_test)
f1 = f1_score(y_valid, predict_test)

print("Test Data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

Test Data
Accuracy:  0.9736842105263158
F1 score:  0.9793103448275863


Scaling the data (Using MinMax Scaler)

In [112]:
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

load = load_breast_cancer()
cancer = pd.DataFrame(load.data, columns=load.feature_names)
target = pd.DataFrame(load.target, columns =['target'])

# Split 80 : 20
X_train, X_valid, y_train, y_valid = train_test_split(cancer, target, train_size = 0.8, test_size = 0.2, random_state=1)

scaler = MinMaxScaler()

X_train = scaler.fit_transform(X_train)
X_valid = scaler.fit_transform(X_valid)


model.fit(X_train,y_train.values.ravel())

LogisticRegression()

In [113]:
predict_train = model.predict(X_train)

acc = accuracy_score(y_train, predict_train)
f1 = f1_score(y_train, predict_train)

print("Local Data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

Local Data
Accuracy:  0.9626373626373627
F1 score:  0.9709401709401708


In [114]:
predict_test = model.predict(X_valid)

acc = accuracy_score(y_valid, predict_test)
f1 = f1_score(y_valid, predict_test)

print("Test Data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

Test Data
Accuracy:  0.9736842105263158
F1 score:  0.9795918367346939


#### Kesimpulan

no scaler: (convergence warning)
- Accuracy:  0.9473684210526315
- F1 score:  0.9589041095890412

with scaling 
- standard
    - Accuracy:  0.9736842105263158
    - F1 score:  0.9793103448275863
- minmax
    - Accuracy:  0.9736842105263158
    - F1 score:  0.9795918367346939
    
This data is better with scaling.

### Neutral_network

### SVM

Define the SVM (Support Vector Machines) constructor

In [115]:
from sklearn.svm import SVC

svm = SVC(random_state=1)

Train the traning dataset

In [116]:
svm.fit(X_train, y_train.values.ravel())
# .values will give the values in an array (shape: (n,1)
# .ravel will convert that array shape to (n, )

SVC(random_state=1)

Predict the dataset.

In [117]:
pred_train_svm_breast = svm.predict(X_train)
pred_val_svm_breast = svm.predict(X_valid)
pred_val_svm_breast

array([1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       0, 1, 1, 1])

Evaluate data train prediction with using Accuracy and F1 metrics

In [118]:
acc_score = accuracy_score(y_train, pred_train_svm_breast)
f1 = f1_score(y_train, pred_train_svm_breast)
print("Accuracy : ", acc_score)
print("F1 : ",f1)

Accuracy :  0.9824175824175824
F1 :  0.986013986013986


Evaluate data test prediction with using Accuracy and F1 metrics

In [119]:
acc_score = accuracy_score(y_valid, pred_val_svm_breast)
f1 = f1_score(y_valid, pred_val_svm_breast)
print("Accuracy : ", acc_score)
print("F1 : ",f1)

Accuracy :  0.9736842105263158
F1 :  0.9793103448275863


# Play-Tennis Dataset

Load the dataset first.

In [120]:
tennis = pd.read_csv('datasets/PlayTennis.csv')
X_tennis = tennis.drop(['play'],axis=1)
y_tennis = pd.DataFrame(tennis['play'])

# Split 80 : 20
X_train_tennis, X_valid_tennis, y_train_tennis, y_valid_tennis = train_test_split(X_tennis, y_tennis, train_size = 0.8, test_size = 0.2, random_state=1)
y_valid_tennis

play
3  yes
7   no
6  yes

Check whether there are missing values

In [121]:
cols_with_missing_tennis = [col for col in tennis.columns if tennis[col].isnull().any()]
cols_with_missing_tennis

[]

## Train datasets
### DecisionTreeClassifier

Make the categorical variable to numeric variable

In [122]:
from sklearn.preprocessing import LabelEncoder

cond = (X_train_tennis.dtypes == 'object')
object_cols = list(cond[cond].index)

label_X_train = X_train_tennis.copy()
label_X_valid = X_valid_tennis.copy()

label_encoder = LabelEncoder()
for col in object_cols:
    label_X_train[col] = label_encoder.fit_transform(X_train_tennis[col])
    label_X_valid[col] = label_encoder.transform(X_valid_tennis[col])

# In case needed
label_y_train = label_encoder.fit_transform(y_train_tennis.values.ravel())
label_y_valid = label_encoder.transform(y_valid_tennis.values.ravel())

label_X_train

outlook  temp  humidity  windy
2         0     1         0  False
10        2     2         1   True
4         1     0         1  False
1         2     1         0   True
12        0     1         1  False
0         2     1         0  False
13        1     2         0   True
9         1     2         1  False
8         2     0         1  False
11        0     2         0   True
5         1     0         1   True

Train the model.

In [123]:
tennis_classifier = DecisionTreeClassifier(random_state=1)
tennis_classifier.fit(label_X_train, y_train_tennis)

DecisionTreeClassifier(random_state=1)

Print the tree.

In [124]:
tree_tennis = export_text(tennis_classifier, feature_names = list(X_tennis.columns))
print(tree_tennis)

|--- outlook <= 0.50
|   |--- class: yes
|--- outlook >  0.50
|   |--- humidity <= 0.50
|   |   |--- class: no
|   |--- humidity >  0.50
|   |   |--- windy <= 0.50
|   |   |   |--- class: yes
|   |   |--- windy >  0.50
|   |   |   |--- outlook <= 1.50
|   |   |   |   |--- class: no
|   |   |   |--- outlook >  1.50
|   |   |   |   |--- class: yes



Predict training data with the model.

In [125]:
predict_train_tennis = tennis_classifier.predict(label_X_train)
predict_train_tennis

array(['yes', 'yes', 'yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'yes',
       'no'], dtype=object)

Evaluate the model to training data with Accuracy Metrics

In [126]:
acc_train_tennis = accuracy_score(y_train_tennis,predict_train_tennis)
acc_train_tennis

1.0

Evaluate the model to training data with F1 Metrics

In [127]:
f1_train_tennis = f1_score(y_train_tennis,predict_train_tennis,pos_label='yes')
f1_train_tennis

1.0

Predict the test data using model.

In [128]:
predict_valid_tennis = tennis_classifier.predict(label_X_valid)
predict_valid_tennis

array(['no', 'no', 'yes'], dtype=object)

Evaluate the model with datatest using Accuracy Metrics

In [129]:
acc_valid_tennis = accuracy_score(y_valid_tennis,predict_valid_tennis)
acc_valid_tennis

0.6666666666666666

Evaluate the model with datatest using F1 Metrics

In [130]:
f1_valid_tennis = f1_score(y_valid_tennis,predict_valid_tennis,pos_label='yes')
f1_valid_tennis

0.6666666666666666

### Id3Estimator

In [131]:
id3_model2 = Id3Estimator()
id3_model2.fit(label_X_train, label_y_train)

tree_text2 = export.export_text(id3_model2.tree_, feature_names=list(tennis.columns))
print(tree_text2)


outlook <=0.50: 1 (3) 
outlook >0.50
|   humidity <=0.50: 0 (3) 
|   humidity >0.50
|   |   windy <=0.50: 1 (3) 
|   |   windy >0.50
|   |   |   temp <=1.00: 0 (1) 
|   |   |   temp >1.00: 1 (1) 



Evaluate the model to training data with Accuracy and F1 Metrics

In [132]:
predict_train2 = id3_model2.predict(label_X_train)

print("Accuracy: ", accuracy_score(predict_train2,label_y_train))
print("F1 score: ", f1_score(predict_train2,label_y_train))

Accuracy:  1.0
F1 score:  1.0


Evaluate the model to test data with using Accuracy and F1 Metrics

In [133]:
predict_test2 = id3_model2.predict(label_X_valid)

print("Accuracy: ", accuracy_score(predict_test2,label_y_valid))
print("F1 score: ", f1_score(predict_test2,label_y_valid))

Accuracy:  0.6666666666666666
F1 score:  0.6666666666666666


### KMeans

Define KMeans and then train the dataset.

In [134]:
from sklearn.cluster import KMeans

# Determine how many clusters
n_clst = 2
# Define the KMeans
kmeans_model_2 = KMeans(n_clusters = n_clst, random_state=1)
kmeans_model_2.fit(label_X_train)

KMeans(n_clusters=2, random_state=1)

Predict and produce clusters.

In [135]:
predict_train = kmeans_model_2.predict(label_X_train)
predict_test = kmeans_model_2.predict(label_X_valid)

Evaluate the training data now

In [136]:
# Find out each cluster belong with which class

# First clusterization using first prediction (k_means_train)
# Calculate the accuracy and F1
acc_train = accuracy_score(label_y_train, predict_train)
f1_train = f1_score(label_y_train, predict_train)
print(f"First clusterization accuracy: {acc_train}")
print(f"First clusterization F1: {f1_train}")

# Second clusterization: switch 0 to 1
switched_train = pd.Series(predict_train)
fix = switched_train.apply(switch)
acc_train1= accuracy_score(label_y_train, fix)
f1_train2= f1_score(label_y_train, fix)
print(f"Second clusterization accuracy: {acc_train1}")
print(f"Second clusterization F1: {f1_train2}")

First clusterization accuracy: 0.36363636363636365
First clusterization F1: 0.4615384615384615
Second clusterization accuracy: 0.6363636363636364
Second clusterization F1: 0.6666666666666666


First clusterization is better, so we take the metrics as the final metrics. Evaluate the validation data now

In [137]:
# First clusterization
acc_valid = accuracy_score(label_y_valid, predict_test)
f1_valid = f1_score(label_y_valid, predict_test)
print(f"First clusterization accuracy: {acc_valid}")
print(f"First clusterization F1: {f1_train}")

# Second clusterization: switch 0 to 1
switched_valid = pd.Series(predict_test).apply(switch)
acc_valid_switch = accuracy_score(label_y_valid, switched_valid)
f1_valid_switch = f1_score(label_y_valid, switched_valid)
print(f"Second clusterization accuracy: {acc_valid_switch}")
print(f"Second clusterization F1: {f1_valid_switch}")

First clusterization accuracy: 0.0
First clusterization F1: 0.4615384615384615
Second clusterization accuracy: 1.0
Second clusterization F1: 1.0


Same as when we evaluate with training data, the first clusterization is better and then will be our final metrics score.

### LogisticRegression

Modelling

In [138]:
model = LogisticRegression()

model.fit(label_X_train,label_y_train)

LogisticRegression()

Predict the train data and evaluate the prediction with Accuracy and F1 score.

In [139]:
predict_train = model.predict(label_X_train)

acc = accuracy_score(label_y_train, predict_train)
f1 = f1_score(label_y_train, predict_train)

print("local data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

local data
Accuracy:  0.8181818181818182
F1 score:  0.8750000000000001


In [140]:
predict_test = model.predict(label_X_valid)

acc = accuracy_score(label_y_valid, predict_test)
f1 = f1_score(label_y_valid, predict_test)

print("test data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

test data
Accuracy:  0.6666666666666666
F1 score:  0.8


Scaling : Standard

In [141]:
# Scaling if needed
from sklearn.preprocessing import StandardScaler, MinMaxScaler
scaler = StandardScaler()

label_X_train = scaler.fit_transform(label_X_train)
label_X_valid = scaler.fit_transform(label_X_valid)

model.fit(label_X_train,label_y_train)

LogisticRegression()

In [142]:
predict_train = model.predict(label_X_train)

acc = accuracy_score(label_y_train, predict_train)
f1 = f1_score(label_y_train, predict_train)

print("local data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

local data
Accuracy:  0.9090909090909091
F1 score:  0.9333333333333333


In [143]:
predict_test = model.predict(label_X_valid)

acc = accuracy_score(label_y_valid, predict_test)
f1 = f1_score(label_y_valid, predict_test)

print("test data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

test data
Accuracy:  0.6666666666666666
F1 score:  0.8


Scaling : MinMax

In [144]:
tennis = pd.read_csv('datasets/PlayTennis.csv')
X_tennis = tennis.drop(['play'],axis=1)
y_tennis = pd.DataFrame(tennis['play'])

# Split 80 : 20
X_train_tennis, X_valid_tennis, y_train_tennis, y_valid_tennis = train_test_split(X_tennis, y_tennis, train_size = 0.8, test_size = 0.2, random_state=1)
y_valid_tennis

from sklearn.preprocessing import LabelEncoder

cond = (X_train_tennis.dtypes == 'object')
object_cols = list(cond[cond].index)

label_X_train = X_train_tennis.copy()
label_X_valid = X_valid_tennis.copy()

label_encoder = LabelEncoder()
for col in object_cols:
    label_X_train[col] = label_encoder.fit_transform(X_train_tennis[col])
    label_X_valid[col] = label_encoder.transform(X_valid_tennis[col])

# In case needed
label_y_train = label_encoder.fit_transform(y_train_tennis.values.ravel())
label_y_valid = label_encoder.transform(y_valid_tennis.values.ravel())

In [145]:
predict_train = model.predict(label_X_train)

acc = accuracy_score(label_y_train, predict_train)
f1 = f1_score(label_y_train, predict_train)

print("local data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

local data
Accuracy:  0.8181818181818182
F1 score:  0.8750000000000001


In [146]:
predict_test = model.predict(label_X_valid)

acc = accuracy_score(label_y_valid, predict_test)
f1 = f1_score(label_y_valid, predict_test)

print("test data")
print("Accuracy: ", acc)
print("F1 score: ", f1)

test data
Accuracy:  0.6666666666666666
F1 score:  0.8


SCORE LOCAL DATA

tanpa scaling:
- Accuracy:  0.8181818181818182
- F1 score:  0.8750000000000001

StandardScaler:
- Accuracy:  0.9090909090909091
- F1 score:  0.9333333333333333

MinMaxScaler:
- Accuracy:  0.8181818181818182
- F1 score:  0.8750000000000001

SCORE TEST DATA
- Accuracy:  0.6666666666666666
- F1 score:  0.8

#### Kesimpulan

Local data better with scaling

### Neural_network

### SVM

Define the SVM (Support Vector Machines) constructor

In [147]:
svm = SVC(random_state=1)

Train the dataset

In [148]:
svm.fit(label_X_train, y_train_tennis.values.ravel())
# .values will give the values in an array (shape: (n,1)
# .ravel will convert that array shape to (n, )
# y_train_tennis.values.ravel()

SVC(random_state=1)

Predict the data train and the data test

In [149]:
pred_svm_train_tennis = svm.predict(label_X_train)
pred_svm_val_tennis = svm.predict(label_X_valid)
pred_svm_train_tennis
# pred_svm_val_tennis


array(['yes', 'yes', 'yes', 'no', 'yes', 'no', 'yes', 'yes', 'yes', 'yes',
       'yes'], dtype=object)

Evaluate the prediction of data train with using Accuracy and F1 metrics

In [150]:
acc_score_svm = accuracy_score(y_train_tennis, pred_svm_train_tennis)
# must define the positive label = yes
f1_score_svm = f1_score(y_train_tennis, pred_svm_train_tennis, pos_label="yes")
print("Accuracy : ",acc_score_svm)
print("F1 : ",f1_score_svm)

Accuracy :  0.8181818181818182
F1 :  0.8750000000000001


Evaluate the prediction of data validation with using Accuracy and F1 metrics

In [151]:
acc_score_svm = accuracy_score(y_valid_tennis, pred_svm_val_tennis)
# must define the positive label = yes
f1_score_svm = f1_score(y_valid_tennis, pred_svm_val_tennis, pos_label="yes")
print("Accuracy : ",acc_score_svm)
print("F1 : ",f1_score_svm)

Accuracy :  1.0
F1 :  1.0


# Analysis accuracy and F1 score for each model